pkg/kernel/tools/heal/liveness_check.go

```diff
@@ -68,20 +69,31 @@ func KernelLivenessCheck(deadlineCtx context.Context, conn *networkservice.Conne
 	var aliveCh = make(chan bool)
 	p.OnIdle = func() {
 		aliveCh <- int(atomic.LoadInt32(&count)) == addrCount
-		close(aliveCh)
+		atomic.AddInt32(&count, -count)
 	}
 
-	err := p.Run()
-
-	if err != nil {
-		log.FromContext(deadlineCtx).Error("Ping failed: %s", err.Error())
-		return false
-	}
+	go func() {
+		for i := 0; i < packetCount; i++ {
+			err := p.Run()
+			if err != nil {
+				log.FromContext(deadlineCtx).Error("Ping failed: %s", err.Error())
+			}
+		}
+	}()
```

Comment:
Why do we need the goroutine here?

Reply:
Because `p.Run()` blocks until `p.MaxRTT` timeout and we can't reach `select` at the end of the function until all p.Run() are completed.

Reply:
For this reason we are using `time.Until(deadline) / packetCount`. Are we missing something?

Reply:
We can avoid goroutine if we make `var aliveCh = make(chan bool, packetCount)` instead of `var aliveCh = make(chan bool)`. Unbuffered channel blocks execution of p.OnIdle and the whole function just freezes.

Reply:
I think we can simply make buffered channel.

Reply:
@NikitaSkrynnik Sound good to me :)